contracts/sdk/IpcContract.sol

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.8.19;
+
+import {IpcEnvelope, ReceiptMsg, IpcMsg, IpcMsgKind} from "../src/structs/CrossNet.sol";
+import {EMPTY_BYTES} from "../src/constants/Constants.sol";
+import {IGateway} from "../src/interfaces/IGateway.sol";
+
+// Interface that needs to be implemented by IPC-enabled contracts.
+// This is really convenient to call it from other contracts.
+interface IfaceIpcContract {
+    /// @notice Entrypoint for cross-net messages in IPC-enabled contracts.
+    // solhint-disable-next-line func-name-mixedcase
+    function IpcEntrypoint(IpcEnvelope calldata crossMsg) external payable returns (bytes memory ret);
+}
+
+abstract contract IpcContract {
+    // list of messages in-flight for which the contract
+    // hasn't received an ACK.
+    mapping(bytes32 => IpcEnvelope) public inFlightMsgs;
+    // auxiliary mapping to track the height at which certain
+    // in-flight messages were received.
+    mapping(uint256 => bytes32[]) public inFlightMsgHeights;
+    // last height where there messages were pruned
+    uint256 private lastGcBlock;
+
+    // The adderss of the gateway in the subnet.
+    address public gatewayAddr;
+    /// Number of blocks that the contract will wait for a receipt
+    /// before garbage collecting the in-flight message.
+    uint256 public timeoutBlocks;
+
+    constructor(address gatewayAddr_, uint256 timeoutBlocks_) {
+        gatewayAddr = gatewayAddr_;
+        timeoutBlocks = timeoutBlocks_;
+    }
+
+    error CallerIsNotGateway();
+    error IpcMsgCallFailed();
+    error UnsupportedMsgKind();
+    error UnrecognizedReceipt();
+
+    /// @notice Entrypoint for IPC-enabled contracts. This function is always called by
+    /// the gateway when a `Call` or `Receipt` cross-net messages is targeted to
+    /// a specific address in the subnet.
+    function IpcEntrypoint(IpcEnvelope calldata crossMsg) external returns (bytes memory) {
+        // trigger the garbage collection of in-flight messages
+        // in case any timeout has triggered already.
+        // timeout == 0 means that no in-flight expiraton is set for receipts.
+        if (timeoutBlocks != 0) {
+            inFlightGC();
+        }
+
+        // only the gateway address is allowed to deliver cross-net messages.
+        if (msg.sender != gatewayAddr) {
+            revert CallerIsNotGateway();
+        }
+
+        // internal dispatch of the cross-net message to the right method.
+        if (crossMsg.kind == IpcMsgKind.Call) {
+            IpcMsg memory callMsg = abi.decode(crossMsg.message, (IpcMsg));
+
+            (bool success, bytes memory ret) = address(this).delegatecall(
```

Comment:
Instead of a delegate call, we should just dispatch to an abstract IPC-aware handler method.